Usb configuration

dwkletke opened the pull request:
Gave definitions to the functions in user_config.c that allows the user configurability from a laptop.
-parse_usb_string passes a message and parses it to see if either the debug level or restriction of sensor messages needs to be changed
-max_debug_level returns the max debug level
-allow_sensor_messages returns if sensor messages are allowed

A few comments and a small test were added to the test file (user_config_test.c), and can be removed if need be.

akmorrison committed:
Create user_config.c/h, and unit tests too

All 8 unit tests currently fail. Modify user_config.c until they
pass. Yay test driven development!

user_config is going to be the library that takes strings from the
user over usb and then modifies the configuration, and what the board
prints out when it receives CAN messages

dwkletke committed:
Defined functions: parse_usb_string, max_debug_level, and allow_sensor_messages to allow user configurability from a laptop.
-parse_usb_string passes a message and parses it to see if either the debug level or restriction of sensor messages needs to be changed
-max_debug_level returns the max debug level
-allow_sensor_messages returns if sensor messages are allowed
A few comments and a small test were added to the test file (user_config_test.c), and can be removed if need be.

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,73 @@
+void parse_usb_string(const char *input) {
+	uint8_t iterate = 0;
+	while (input[iterate] != '\0') {
+		switch (change_function) {
```
So you're using change_function to represent the state of this parser. That's really good. The only problem is that the numbers have no semantic meaning, which makes it really hard to read. I think you want an enum type:

Something like this: https://stackoverflow.com/questions/15237656/using-enum-type-in-a-switch-statement

dwkletke replied:
I'm glad you commented on this because I knew that there was a better way but I couldn't come up with one at the time.

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,73 @@
+							printf("\tSensor messages are not allowed!\n");
+					}
+					break;
+				} // @suppress("No break at end of case")
```
Why is there no break here? I'm assuming it's intentional, but why?

dwkletke replied:
There is no break because if the current character being checked is a new message (not apart of the old message being checked), then the function still needs to check it in other parts of the function (i.e. if "S0G1;" is sent into the function, when 'G' is being checked, the function should recognize that "S0G" is not a valid message but should also recognize that 'G' could also be the start of another message). 
I also realized that this check doesn't happen if it is at the second level check (check for the number of the level), so I will change the function accordingly.

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,73 @@
+			// Case 0: new instance of checking for valid message sent
+			// --> if parse_usb_string has received a char (either 'S' or 'G')
+			case 0:
+				if (input[iterate] == 'G') {
```
This is probably a good time to use a switch statement. It'll save you a couple lines, and it'll be easier to add new commands in the future, if we find any.

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,73 @@
+				}
+				else if (input[iterate] == 'L') {
+					save_type = 'L';
+					change_function += 2;
```
Yeah, please don't use increments to change_function, just assign it to the value you want

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,73 @@
+					change_function = 0;
+				break;
+		}
+		iterate++;
```
I know a lot of people say that variable names should be expressive and whatnot, but naming your iteration variable `i` is a very standard pattern in C. Please just name this `i`

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,73 @@
+						printf("Current Config: \n");
+						printf("\tMax debug level: %d\n", max_debug_level());
+						if (allow_sensor_messages())
+							printf("\tSensor messages are allowed!\n");
```
Can you use whatever the print function we use to print over usb is in usb_app.c? I don't actually remember what it's called.

you may need to use `sprintf` to get the numbers in there first.

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,73 @@
+#include "user_config.h"
+#include <stdio.h>
+
+static uint8_t set_debug_level = 0, change_function = 0, temp_num = 0;
```
Please don't declare multiple variables on the same line. I know I've been guilty of doing it before, but it's much more readable with one per line

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,73 @@
+
+static uint8_t set_debug_level = 0, change_function = 0, temp_num = 0;
+static bool set_sensor_messages = false;
+static char save_type;
```
What does this variable do? Can you leave a quick comment?

dwkletke replied:
save_type is the variable used to save what kind of message is being sent (either 'G' or 'S'). I will add a comment explaining this!

akmorrison reviewed (changes requested):
All the tests pass on my machine, good job. All changes are stylistic and for readability

dwkletke committed:
Addressed requested changes:
-Made the level check switch statement check for a enum representing specific levels of parsing
-Added switch statements for code readability and ease of use
-Changed name of iterate to i
-Changed the print config functionality to rely on print function in usb_app.c
-Changed declarations of variables to be on separate lines
-Added some comments to explain variable meaning

dwkletke committed:
-Changed the string length sent into usb_app_write_string

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,90 @@
+					sprintf(config_msg, "Current Config: Max debug level = %d & Sensor messages %s allowed!", max_debug_level(), sensor_check);
+					config_msg[71] = '\n';
+					config_msg[72] = '\0';
+					usb_app_write_string(config_msg, 72);
```
No need to use a hardcoded 72 value here, sprintf returns the number of characters that it wrote into the buffer (http://www.cplusplus.com/reference/cstdio/sprintf/)

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,90 @@
+					strcpy(sensor_check, "are");
+					if (!allow_sensor_messages())
+						strcat(sensor_check, " not");
+					sprintf(config_msg, "Current Config: Max debug level = %d & Sensor messages %s allowed!", max_debug_level(), sensor_check);
```
I highly recommend using `snprintf` here, since it guarantees that it won't write more characters than you let it

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,90 @@
+					if (!allow_sensor_messages())
+						strcat(sensor_check, " not");
+					sprintf(config_msg, "Current Config: Max debug level = %d & Sensor messages %s allowed!", max_debug_level(), sensor_check);
+					config_msg[71] = '\n';
```
Due to legacy serial reasons that aren't important, you actually want "\r\n" at the end of lines instead of just "\n". Blog post if you're curious about why:
https://blog.codinghorror.com/the-great-newline-schism/

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,90 @@
+static char save_type;
+
+// function takes in a message and parses it to see if max_debug_level() or allow_sensor_messages() needs to be changed
+void parse_usb_string(const char *input) {
```
Can you add a quick check at the top of this function that if `input == NULL`, you return early?

No one should ever call this with a null value, but it doesn't hurt to double check. Generally in C it's good practice to always check the pointer values in your function

akmorrison reviewed (changes requested):
4 small changes

dwkletke committed:
Addressed more requested changes:
-Used snprintf to save the size of the message (and be used for usb_app_write_string)
-Used snprintf instead of sprintf to limit the size of the message
-Used "\r\n" instead of "\n"
-Added check for if input message pointer was not equal to NULL --> then let input message go through function
	-If input message is false, exit function early

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,96 @@
+						if (!allow_sensor_messages())
+							strcat(sensor_check, " not");
+						msg_length = snprintf(config_msg, 72, "Current Config: Max debug level = %d & Sensor messages %s allowed!", max_debug_level(), sensor_check);
+						config_msg[71] = '\r\n';
```
use msg_length here instead of 71

akmorrison replied:
Also, \r\n isn't a single character, it's two characters in a row. Highly recommend putting this in the snprintf call

akmorrison commented on user_config.c:
```diff
@@ -0,0 +1,96 @@
+						msg_length = snprintf(config_msg, 72, "Current Config: Max debug level = %d & Sensor messages %s allowed!", max_debug_level(), sensor_check);
+						config_msg[71] = '\r\n';
+						config_msg[72] = '\0';
+						usb_app_write_string(config_msg, msg_length);
```
check the docs to see if msg_length includes the null terminator, I'm not sure if it does

dwkletke replied:
It looks like snprintf includes the null terminator as long as the length is not 0 (which it should never be in our case since something is always being written into the string).

akmorrison replied:
ok, so if snprintf includes that and it's writing it to the config_msg, you shouldn't need the line
`config_msg[72] = '\0'`

dwkletke replied:
Yeah I took that out!

dwkletke committed:
Addressed a few tweaks:
-Put formatting (\r\n) into snprintf line defining config_msg
-Took away definition of null terminator (since snprintf already includes \0)

akmorrison committed:
Call user_config code on inputted strings

Minor modifications to make it run on picdev hardware. Also, I can
confirm that this code runs on hardware, and it works perfectly
according to my cursory testing.
